### PR TITLE
WIP: CLE Filters

### PR DESCRIPTION
**Pull request by aaron-bray**

### CODE REVIEW

**Comment by aaron-bray** [comment]
@tbirdso I am not sure how these python packages are being built, but they seem to be failing to fing the #include "clic.hpp" file, which is generated at CLIc cmake configuration and placed in the binary directory. CMake is not failing... so I assume the compiler (or scikit-build) is not being told to add the CLIc binary directory to its include path? I am not sure how all this automation is working...

**Comment by tbirdso** [comment]
Hi @aaron-bray , just getting back from vacation. I agree with your analysis. It looks like the CLEsperanto `development` branch makes `clic.hpp` configurable via CMake, so adding the build directory `include` folder to ITKCLEsperanto's include paths seems to have fixed the issue on my local system. Re-running CI now.

In the long term it would be nice to reference a stable `CLIc_prototype` tag rather than `master` or `development` branches, but fine to stick with new features in the `development` branch for now.

There is one compiler warning in CLIc tests that results in a CI failure, best path forward is probably to turn off building CLIc_prototype tests in ITKCLEsperanto CI.

**Comment by tbirdso** [comment]
@aaron-bray CI is green after disabling `CLIc_prototype` tests 👍

File: CMakeLists.txt [hunk 1]
```diff
@@ -22,16 +22,23 @@ macro(get_all_targets_recursive targets dir)
...
+# Enable/disable ITK/CLIc test suites independently.
+# Note that clci inherits the value of BUILD_TESTING when clci is declared.
+set(ITKCLEsperanto_BUILD_TESTING ${BUILD_TESTING})
+option(CLIc_BUILD_TESTING OFF)
+set(BUILD_TESTING ${CLIc_BUILD_TESTING})
+
 # Get CLIc
 set(CLIc_GIT_REPOSITORY "https://github.com/clEsperanto/CLIc_prototype.git")
-set(CLIc_GIT_TAG        "master")
+set(CLIc_GIT_TAG        "development")
```
Comment (thewtex) [inline CMakeLists.txt:41]: Let's use a specific hash as suggested by @tbirdso and make a comment noting the branch and data the commit comes from.

File: include/itkCLEAddImageFilter.h [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+#ifndef itkCLEAddImageFilter_h
+#define itkCLEAddImageFilter_h
+
+#include "itkAddImageFilter.h"
+#include "itkCLEBinaryGeneratorImageFilter.h"
+
+namespace itk
+{
+/** \class CLEAddImageFilter
+ * \brief Base class for CLE filters that take an image as input and overwrite that image as the output
```
Comment (thewtex) [inline include/itkCLEAddImageFilter.h:27]: Needs update

File: include/itkCLEAddImageFilter.h [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+ *
+ * This class is the base class for CLE inplace filter. The template parameter for parent class type
```
Comment (thewtex) [inline include/itkCLEAddImageFilter.h:29]: Needs update

File: include/itkCLEImage.hxx [hunk 4]
```diff
@@ -49,6 +49,16 @@ CLEImage<TPixel, VImageDimension>::Allocate(bool initialize)
   this->m_DataManager->SetTimeStamp(this->GetTimeStamp());
 }
 
+template <typename TPixel, unsigned int VImageDimension>
+void
+CLEImage<TPixel, VImageDimension>::AllocateGPU(bool initialize)
+{
+  this->m_DataManager->AllocateGPU();
```
Comment (thewtex) [inline include/itkCLEImage.hxx:56]: Can we implement `initialize`, i.e. set all buffer values to zero if `true`?

File: CMakeLists.txt (hunk 1 above)
Comment (tbirdso) [reply CMakeLists.txt]: I've gone ahead and set a hash at the current `development` branch head to mitigate future CI issues.

File: test/CMakeLists.txt [hunk 5]
```diff
@@ -4,6 +4,7 @@ message(STATUS "Adding Tests")
 
 set(CLEsperantoTests
   itkCLEImageTest.cxx
+  itkCLEAddImageTest.cxx
```
Comment (tbirdso) [inline test/CMakeLists.txt:7]: In addition to adding the source file reference here, another test entry needs to be added with `itk_add_test` (see example on line 12 below).

**Comment by aaron-bray** [comment]
@thewtex Please have another review

**Review by thewtex — APPROVED** [review]
@aaron-bray @tbirdso awesome work!

**Pull request merged**